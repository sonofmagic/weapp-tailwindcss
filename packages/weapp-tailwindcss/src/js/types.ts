import type { NodePath } from '@babel/traverse'
import type { StringLiteral, TemplateElement } from '@babel/types'

export interface JsTokenMeta {
  ignore?: boolean
}

// interface JsTokenTemplateElement {
//   type: 'TemplateElement'
//   ast: TemplateElement
// }

// interface JsTokenStringLiteral {
//   type: 'StringLiteral'
//   ast: StringLiteral
// }

//  (JsTokenTemplateElement | JsTokenStringLiteral) &

/**
 * Represents a slice of the original source that must be updated while keeping
 * track of the owning AST path for later filtering.
 */
export interface JsToken {
  start: number
  end: number
  value: string
  // type?: (string & {})
  path: NodePath<StringLiteral | TemplateElement>
}
