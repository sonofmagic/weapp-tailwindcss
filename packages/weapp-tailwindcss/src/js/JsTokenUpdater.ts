import type MagicString from 'magic-string'
import type { JsToken } from './types'

/**
 * Lightweight helper that batches updates to {@link MagicString}.
 * It keeps the transformation logic out of the traversal code and makes
 * it easier to reason about the order in which tokens are written back.
 */
export class JsTokenUpdater {
  private tokens: JsToken[]

  constructor({ value }: { value?: JsToken[] } = {}) {
    this.tokens = value ? [...value] : []
  }

  addToken(token?: JsToken) {
    if (token) {
      this.tokens.push(token)
    }
  }

  push(...args: JsToken[]) {
    this.tokens.push(...args)
    return this
  }

  map(callbackfn: (value: JsToken, index: number, array: JsToken[]) => JsToken) {
    this.tokens = this.tokens.map(callbackfn)
    return this
  }

  filter(callbackfn: (value: JsToken, index: number, array: JsToken[]) => unknown) {
    this.tokens = this.tokens.filter(callbackfn)
    return this
  }

  updateMagicString(ms: MagicString) {
    for (const { start, end, value } of this.tokens) {
      ms.update(start, end, value)
    }
    return ms
  }
}
