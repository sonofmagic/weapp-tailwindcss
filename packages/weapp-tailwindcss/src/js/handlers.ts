// import type { StringLiteral, TemplateElement, Comment } from '@babel/types'
import type MagicString from 'magic-string'
import type { IJsHandlerOptions } from '../types'
import { replaceWxml } from '../wxml/shared'
import { escapeStringRegexp } from '../reg'
import { splitCode } from '../extractors/split'
import { decodeUnicode2, jsStringEscape } from '../escape'
// https://developer.mozilla.org/zh-CN/docs/Web/JavaScript/Reference/Global_Objects/String

interface ReplaceNode {
  leadingComments?: { value: string }[] | null | undefined
  start?: number | null
  end?: number | null
}

export function toUnicodeEscapedString(str: string) {
  return str.split('').map((char) => {
    const code = char.charCodeAt(0)
    if (code <= 31) {
      const hexCode = code.toString(16).padStart(4, '0')
      return `\\u${hexCode}`
    }
    return char
  }).join('')
}

export function toUnicodeEscapedString2(str: string) {
  return escape(str).replace(/%u/g, '\\u').replace(/%/g, '\\u00')
}

export function replaceHandleValue(
  str: string,
  node: ReplaceNode,
  options: IJsHandlerOptions,
  ms: MagicString,
  offset = 0,
) {
  const {
    classNameSet: set,
    escapeMap,
    mangleContext: ctx,
    needEscaped = false,
    jsPreserveClass,
    arbitraryValues,
    always,
    unescapeUnicode,
  } = options

  const allowDoubleQuotes = arbitraryValues?.allowDoubleQuotes

  let rawStr = str
  let needUpdate = false
  if (unescapeUnicode && rawStr.includes('\\u')) {
    rawStr = decodeUnicode2(rawStr)
  }
  const arr = splitCode(rawStr, allowDoubleQuotes)
  for (const v of arr) {
    if (always || (set && set.has(v) && !jsPreserveClass?.(v))) {
      let ignoreFlag = false
      if (Array.isArray(node.leadingComments)) {
        ignoreFlag
          = node.leadingComments.findIndex(x => x.value.includes('weapp-tw') && x.value.includes('ignore')) > -1
      }

      if (!ignoreFlag) {
        if (ctx) {
          rawStr = ctx.jsHandler(rawStr)
        }

        rawStr = rawStr.replace(
          new RegExp(escapeStringRegexp(v)),
          replaceWxml(v, {
            escapeMap,
          }),
        )
        needUpdate = true
      }
    }
  }

  if (needUpdate && typeof node.start === 'number' && typeof node.end === 'number') {
    const start = node.start + offset
    const end = node.end - offset

    if (start < end && str !== rawStr) {
      const content = needEscaped ? jsStringEscape(rawStr) : rawStr
      ms.update(start, end, content)
    }
  }

  return rawStr
}
