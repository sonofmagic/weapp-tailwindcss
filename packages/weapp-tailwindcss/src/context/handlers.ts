import type { IMangleScopeContext } from '@weapp-tailwindcss/mangle'
import type { ICustomAttributesEntities, InternalUserDefinedOptions } from '@/types'
import { createStyleHandler } from '@weapp-tailwindcss/postcss'
import { createJsHandler } from '@/js'
import { createTemplateHandler } from '@/wxml'

export function createHandlersFromContext(
  ctx: InternalUserDefinedOptions,
  mangleContext: IMangleScopeContext,
  customAttributesEntities: ICustomAttributesEntities,
  cssCalcOptions: InternalUserDefinedOptions['cssCalc'],
) {
  const {
    cssPreflight,
    customRuleCallback,
    cssPreflightRange,
    escapeMap,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    px2rpx,
    arbitraryValues,
    jsPreserveClass,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    inlineWxs,
    disabledDefaultTemplateHandler,
  } = ctx

  const styleHandler = createStyleHandler({
    cssPreflight,
    customRuleCallback,
    cssPreflightRange,
    escapeMap,
    mangleContext,
    cssChildCombinatorReplaceValue,
    injectAdditionalCssVarScope,
    cssSelectorReplacement,
    rem2rpx,
    postcssOptions,
    cssRemoveProperty,
    cssRemoveHoverPseudoClass,
    cssPresetEnv,
    uniAppX,
    cssCalc: cssCalcOptions,
    px2rpx,
  })

  const jsHandler = createJsHandler({
    escapeMap,
    mangleContext,
    arbitraryValues,
    jsPreserveClass,
    generateMap: true,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
  })

  const templateHandler = createTemplateHandler({
    customAttributesEntities,
    escapeMap,
    mangleContext,
    inlineWxs,
    jsHandler,
    disabledDefaultTemplateHandler,
  })

  return {
    styleHandler,
    jsHandler,
    templateHandler,
  }
}
