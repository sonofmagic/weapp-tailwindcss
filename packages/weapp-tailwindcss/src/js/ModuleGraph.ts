import type { ParserOptions } from '@babel/parser'
import type { NodePath } from '@babel/traverse'
import type { Node as BabelNode, Identifier, StringLiteral } from '@babel/types'
import type { IJsHandlerOptions, JsModuleGraphOptions, LinkedJsModuleResult } from '../types'
import type { SourceAnalysis } from './babel'
import type { ImportToken } from './NodePathWalker'
import { analyzeSource, babelParse, processUpdatedSource } from './babel'

interface ModuleState {
  filename: string
  source: string
  analysis: SourceAnalysis
}

interface QueueItem {
  filename: string
  depth: number
}

export interface ModuleGraphEntry {
  filename: string
  source: string
  analysis: SourceAnalysis
  handlerOptions: IJsHandlerOptions
}

export class JsModuleGraph {
  private readonly modules = new Map<string, ModuleState>()
  private readonly queue: QueueItem[] = []
  private readonly ignoredExportNames = new Map<string, Set<string>>()
  private readonly resolve: JsModuleGraphOptions['resolve']
  private readonly load: JsModuleGraphOptions['load']
  private readonly filter?: JsModuleGraphOptions['filter']
  private readonly maxDepth: number
  private readonly baseOptions: IJsHandlerOptions
  private readonly parserOptions?: ParserOptions
  private readonly rootFilename: string

  constructor(entry: ModuleGraphEntry, graphOptions: JsModuleGraphOptions) {
    this.resolve = graphOptions.resolve
    this.load = graphOptions.load
    this.filter = graphOptions.filter
    this.maxDepth = graphOptions.maxDepth ?? Number.POSITIVE_INFINITY
    const { moduleGraph: _moduleGraph, filename: _ignoredFilename, ...rest } = entry.handlerOptions
    this.baseOptions = {
      ...rest,
      filename: entry.filename,
    }
    this.parserOptions = entry.handlerOptions.babelParserOptions
    this.rootFilename = entry.filename

    this.modules.set(entry.filename, {
      filename: entry.filename,
      source: entry.source,
      analysis: entry.analysis,
    })
    this.queue.push({ filename: entry.filename, depth: 0 })
  }

  build(): Record<string, LinkedJsModuleResult> {
    this.collectDependencies()

    const linked: Record<string, LinkedJsModuleResult> = {}
    for (const [filename, state] of this.modules) {
      if (filename === this.rootFilename) {
        continue
      }

      const childOptions: IJsHandlerOptions = {
        ...this.baseOptions,
        filename,
      }
      const ms = processUpdatedSource(state.source, childOptions, state.analysis)
      const code = ms.toString()
      if (code !== state.source) {
        linked[filename] = { code }
      }
    }

    return linked
  }

  private addIgnoredExport(filename: string, exportName: string) {
    if (!exportName) {
      return
    }
    let pending = this.ignoredExportNames.get(filename)
    if (!pending) {
      pending = new Set()
      this.ignoredExportNames.set(filename, pending)
    }
    if (pending.has(exportName)) {
      return
    }
    pending.add(exportName)
    const existing = this.modules.get(filename)
    if (existing) {
      this.applyIgnoredExportsToAnalysis(filename, existing.analysis)
    }
  }

  private registerIgnoredExportsFromTokens(resolved: string, tokens: ImportToken[]) {
    for (const token of tokens) {
      if (token.type === 'ImportSpecifier') {
        this.addIgnoredExport(resolved, token.imported)
      }
      else if (token.type === 'ImportDefaultSpecifier') {
        this.addIgnoredExport(resolved, 'default')
      }
    }
  }

  private applyIgnoredExportsToAnalysis(filename: string, analysis: SourceAnalysis) {
    const pending = this.ignoredExportNames.get(filename)
    if (!pending || pending.size === 0) {
      return
    }

    const names = new Set(pending)
    pending.clear()

    const propagate: Array<{ specifier: string, exportName: string }> = []

    for (const exportPath of analysis.exportDeclarations) {
      if (names.size === 0) {
        break
      }

      if (exportPath.isExportDefaultDeclaration()) {
        if (names.has('default')) {
          analysis.walker.walkExportDefaultDeclaration(exportPath)
          names.delete('default')
        }
        continue
      }

      if (exportPath.isExportNamedDeclaration()) {
        const source = exportPath.node.source?.value
        if (typeof source === 'string') {
          for (const spec of exportPath.get('specifiers')) {
            if (!spec.isExportSpecifier()) {
              continue
            }
            const exported = spec.get('exported')
            let exportedName: string | undefined
            if (exported.isIdentifier()) {
              exportedName = exported.node.name
            }
            else if (exported.isStringLiteral()) {
              exportedName = exported.node.value
            }

            if (!exportedName || !names.has(exportedName)) {
              continue
            }

            const local = spec.get('local') as NodePath<Identifier | StringLiteral>
            if (local.isIdentifier()) {
              propagate.push({
                specifier: source,
                exportName: local.node.name,
              })
              names.delete(exportedName)
            }
            else if (local.isStringLiteral()) {
              propagate.push({
                specifier: source,
                exportName: local.node.value,
              })
              names.delete(exportedName)
            }
          }
          continue
        }

        const declaration = exportPath.get('declaration')
        if (declaration.isVariableDeclaration()) {
          for (const decl of declaration.get('declarations')) {
            const id = decl.get('id')
            if (id.isIdentifier()) {
              const exportName = id.node.name
              if (names.has(exportName)) {
                analysis.walker.walkVariableDeclarator(decl)
                names.delete(exportName)
              }
            }
          }
        }

        for (const spec of exportPath.get('specifiers')) {
          if (!spec.isExportSpecifier()) {
            continue
          }
          const exported = spec.get('exported')
          let exportedName: string | undefined
          if (exported.isIdentifier()) {
            exportedName = exported.node.name
          }
          else if (exported.isStringLiteral()) {
            exportedName = exported.node.value
          }
          if (!exportedName || !names.has(exportedName)) {
            continue
          }
          const local = spec.get('local') as NodePath<BabelNode | null | undefined>
          analysis.walker.walkNode(local)
          names.delete(exportedName)
        }
        continue
      }

      if (exportPath.isExportAllDeclaration()) {
        const source = exportPath.node.source?.value
        if (typeof source === 'string') {
          for (const exportName of names) {
            propagate.push({
              specifier: source,
              exportName,
            })
          }
          names.clear()
        }
      }
    }

    for (const { specifier, exportName } of propagate) {
      let resolved: string | undefined
      try {
        resolved = this.resolve(specifier, filename)
      }
      catch {
        resolved = undefined
      }
      if (!resolved) {
        pending.add(exportName)
        continue
      }
      if (this.filter && !this.filter(resolved, specifier, filename)) {
        pending.add(exportName)
        continue
      }
      this.addIgnoredExport(resolved, exportName)
    }

    for (const name of names) {
      pending.add(name)
    }
  }

  private collectDependencies() {
    while (this.queue.length > 0) {
      const { filename, depth } = this.queue.shift() as QueueItem
      if (depth >= this.maxDepth) {
        continue
      }
      const state = this.modules.get(filename)
      if (!state) {
        continue
      }

      const dependencySpecifiers = new Map<string, ImportToken[]>()
      for (const token of state.analysis.walker.imports) {
        if (!dependencySpecifiers.has(token.source)) {
          dependencySpecifiers.set(token.source, [])
        }
        dependencySpecifiers.get(token.source)!.push(token)
      }
      for (const exportPath of state.analysis.exportDeclarations) {
        if (exportPath.isExportAllDeclaration() || exportPath.isExportNamedDeclaration()) {
          const source = exportPath.node.source?.value
          if (typeof source === 'string' && !dependencySpecifiers.has(source)) {
            dependencySpecifiers.set(source, [])
          }
        }
      }

      for (const [specifier, tokens] of dependencySpecifiers) {
        let resolved: string | undefined
        try {
          resolved = this.resolve(specifier, filename)
        }
        catch {
          continue
        }
        if (!resolved) {
          continue
        }
        if (this.filter && !this.filter(resolved, specifier, filename)) {
          continue
        }
        if (tokens.length > 0) {
          this.registerIgnoredExportsFromTokens(resolved, tokens)
        }
        if (this.modules.has(resolved)) {
          continue
        }

        let source: string | undefined
        try {
          source = this.load(resolved)
        }
        catch {
          continue
        }
        if (typeof source !== 'string') {
          continue
        }

        let analysis: SourceAnalysis
        try {
          const ast = babelParse(source, {
            ...this.parserOptions,
            sourceFilename: resolved,
          })
          analysis = analyzeSource(ast, {
            ...this.baseOptions,
            filename: resolved,
          })
          this.applyIgnoredExportsToAnalysis(resolved, analysis)
        }
        catch {
          continue
        }

        this.modules.set(resolved, {
          filename: resolved,
          source,
          analysis,
        })
        this.queue.push({ filename: resolved, depth: depth + 1 })
      }
    }
  }
}
