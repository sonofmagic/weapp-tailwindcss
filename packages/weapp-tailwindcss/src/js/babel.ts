import type { ParseError, ParseResult, ParserOptions } from '@babel/parser'
import type { NodePath, TraverseOptions } from '@babel/traverse'
import type { CallExpression, ExportDeclaration, File, ImportDeclaration, Node, StringLiteral, TemplateElement } from '@babel/types'
import type { IJsHandlerOptions, JsHandlerResult } from '../types'
import type { JsToken } from './types'
import { jsStringEscape } from '@ast-core/escape'
import { LRUCache } from 'lru-cache'
import MagicString from 'magic-string'
import { parse, traverse } from '@/babel'
import { createNameMatcher } from '@/utils/nameMatcher'
import { replaceHandleValue } from './handlers'
import { JsTokenUpdater } from './JsTokenUpdater'
import { JsModuleGraph } from './ModuleGraph'
import { NodePathWalker } from './NodePathWalker'

/**
 * Describes the data collected during AST traversal that is required for the
 * subsequent source transformation phase.
 */
export interface SourceAnalysis {
  ast: ParseResult<File>
  walker: NodePathWalker
  jsTokenUpdater: JsTokenUpdater
  targetPaths: NodePath<StringLiteral | TemplateElement>[]
  importDeclarations: Set<NodePath<ImportDeclaration>>
  exportDeclarations: Set<NodePath<ExportDeclaration>>
  requireCallPaths: NodePath<StringLiteral>[]
  ignoredPaths: WeakSet<NodePath<StringLiteral | TemplateElement>>
}

type EvalHandler = (source: string, opts: IJsHandlerOptions) => JsHandlerResult

export const parseCache: LRUCache<string, ParseResult<File>> = new LRUCache<string, ParseResult<File>>(
  {
    max: 512,
  },
)

export function genCacheKey(source: string, options: any): string {
  return (
    source
    + JSON.stringify(options, (_, val) =>
      typeof val === 'function' ? val.toString() : val)
  )
}

export function babelParse(
  code: string,
  { cache, ...options }: ParserOptions & { cache?: boolean } = {},
) {
  const cacheKey = genCacheKey(code, options)
  let result: ParseResult<File> | undefined
  if (cache) {
    result = parseCache.get(cacheKey)
  }

  if (!result) {
    result = parse(code, options)
    if (cache) {
      parseCache.set(cacheKey, result)
    }
  }

  return result
}
export function isEvalPath(p: NodePath<Node>) {
  if (p.isCallExpression()) {
    const calleePath = p.get('callee')
    return calleePath.isIdentifier(
      {
        name: 'eval',
      },
    )
  }
  return false
}

/**
 * Normalises replacement metadata for literals discovered inside `eval` calls.
 */
function createEvalReplacementToken(
  path: NodePath<StringLiteral | TemplateElement>,
  updated: string,
): JsToken | undefined {
  const node = path.node

  let offset = 0
  let original: string
  if (path.isStringLiteral()) {
    offset = 1
    original = path.node.value
  }
  else if (path.isTemplateElement()) {
    original = path.node.value.raw
  }
  else {
    original = ''
  }

  if (typeof node.start !== 'number' || typeof node.end !== 'number') {
    return undefined
  }

  const start = node.start + offset
  const end = node.end - offset
  if (start >= end) {
    return undefined
  }

  if (original === updated) {
    return undefined
  }

  const value = path.isStringLiteral() ? jsStringEscape(updated) : updated

  return {
    start,
    end,
    value,
    path,
  }
}

function createModuleSpecifierReplacementToken(
  path: NodePath<StringLiteral>,
  replacement: string,
): JsToken | undefined {
  const node = path.node
  if (node.value === replacement) {
    return undefined
  }

  if (typeof node.start !== 'number' || typeof node.end !== 'number') {
    return undefined
  }

  const start = node.start + 1
  const end = node.end - 1
  if (start >= end) {
    return undefined
  }

  return {
    start,
    end,
    value: replacement,
    path,
  }
}

function collectModuleSpecifierReplacementTokens(
  analysis: SourceAnalysis,
  replacements: Record<string, string>,
) {
  const tokens: JsToken[] = []

  const applyReplacement = (path: NodePath<StringLiteral>) => {
    const replacement = replacements[path.node.value]
    if (!replacement) {
      return
    }
    const token = createModuleSpecifierReplacementToken(path, replacement)
    if (token) {
      tokens.push(token)
    }
  }

  for (const importPath of analysis.importDeclarations) {
    const source = importPath.get('source')
    if (source.isStringLiteral()) {
      applyReplacement(source)
    }
  }

  for (const exportPath of analysis.exportDeclarations) {
    if (exportPath.isExportNamedDeclaration() || exportPath.isExportAllDeclaration()) {
      const source = exportPath.get('source')
      if (source && !Array.isArray(source) && source.isStringLiteral()) {
        applyReplacement(source)
      }
    }
  }

  for (const literalPath of analysis.requireCallPaths) {
    applyReplacement(literalPath)
  }

  for (const token of analysis.walker.imports) {
    const replacement = replacements[token.source]
    if (replacement) {
      token.source = replacement
    }
  }

  return tokens
}

function handleEvalStringLiteral(
  path: NodePath<StringLiteral>,
  options: IJsHandlerOptions,
  updater: JsTokenUpdater,
  handler: EvalHandler,
) {
  const { code } = handler(path.node.value, {
    ...options,
    needEscaped: false,
    generateMap: false,
  })

  if (!code) {
    return
  }

  const token = createEvalReplacementToken(path, code)
  if (token) {
    updater.addToken(token)
  }
}

function handleEvalTemplateElement(
  path: NodePath<TemplateElement>,
  options: IJsHandlerOptions,
  updater: JsTokenUpdater,
  handler: EvalHandler,
) {
  const { code } = handler(path.node.value.raw, {
    ...options,
    generateMap: false,
  })

  if (!code) {
    return
  }

  const token = createEvalReplacementToken(path, code)
  if (token) {
    updater.addToken(token)
  }
}

/**
 * Recursively analyse string-like nodes that sit inside `eval` calls so that we
 * can rewrite them while keeping their offsets intact.
 */
function walkEvalExpression(
  path: NodePath<CallExpression>,
  options: IJsHandlerOptions,
  updater: JsTokenUpdater,
  handler: EvalHandler,
) {
  path.traverse({
    StringLiteral(innerPath) {
      handleEvalStringLiteral(innerPath, options, updater, handler)
    },
    TemplateElement(innerPath) {
      handleEvalTemplateElement(innerPath, options, updater, handler)
    },
  })
}

export function analyzeSource(
  ast: ParseResult<File>,
  options: IJsHandlerOptions,
  handler?: EvalHandler,
): SourceAnalysis {
  const jsTokenUpdater = new JsTokenUpdater()
  const ignoredPaths = new WeakSet<NodePath<StringLiteral | TemplateElement>>()
  const walker = new NodePathWalker(
    {
      ignoreCallExpressionIdentifiers: options.ignoreCallExpressionIdentifiers,
      callback(path) {
        if (path.isStringLiteral() || path.isTemplateElement()) {
          ignoredPaths.add(path)
        }
      },
    },
  )

  const isIgnoredTaggedTemplate = createNameMatcher(options.ignoreTaggedTemplateExpressionIdentifiers, { exact: true })

  const targetPaths: NodePath<StringLiteral | TemplateElement>[] = []
  const importDeclarations = new Set<NodePath<ImportDeclaration>>()
  const exportDeclarations = new Set<NodePath<ExportDeclaration>>()
  const requireCallPaths: NodePath<StringLiteral>[] = []
  // eslint-disable-next-line ts/no-use-before-define -- default handler references exported jsHandler defined later
  const evalHandler = handler ?? jsHandler

  const traverseOptions: TraverseOptions<Node> = {
    StringLiteral: {
      enter(p) {
        if (isEvalPath(p.parentPath)) {
          return
        }
        targetPaths.push(p)
      },
    },
    TemplateElement: {
      enter(p) {
        const pp = p.parentPath
        if (pp.isTemplateLiteral()) {
          const ppp = pp.parentPath
          if (isEvalPath(ppp)) {
            return
          }
          if (ppp.isTaggedTemplateExpression()) {
            const tagPath = ppp.get('tag')
            if (
              (tagPath.isIdentifier()
                && isIgnoredTaggedTemplate(tagPath.node.name)
              )
            ) {
              return
            }
          }
        }
        targetPaths.push(p)
      },
    },
    CallExpression: {
      enter(p) {
        if (isEvalPath(p)) {
          walkEvalExpression(p, options, jsTokenUpdater, evalHandler)
          return
        }

        const calleePath = p.get('callee')
        if (
          calleePath.isIdentifier({ name: 'require' })
          && !p.scope.hasBinding('require')
        ) {
          const args = p.get('arguments')
          if (Array.isArray(args) && args.length > 0) {
            const first = args[0]
            if (first?.isStringLiteral()) {
              requireCallPaths.push(first)
            }
          }
        }

        walker.walkCallExpression(p)
      },
    },
    ImportDeclaration: {
      enter(p) {
        importDeclarations.add(p)
      },
    },
    ExportDeclaration: {
      enter(p) {
        exportDeclarations.add(p)
      },
    },
  }

  traverse(ast, traverseOptions)

  return {
    walker,
    jsTokenUpdater,
    ast,
    targetPaths,
    importDeclarations,
    exportDeclarations,
    requireCallPaths,
    ignoredPaths,
  }
}

export function processUpdatedSource(
  rawSource: string,
  options: IJsHandlerOptions,
  analysis: SourceAnalysis,
) {
  const ms = new MagicString(rawSource)
  const { targetPaths, jsTokenUpdater, ignoredPaths } = analysis

  // Build replacement tokens for every string-like node collected earlier.
  const replacementTokens: JsToken[] = []
  for (const path of targetPaths) {
    if (ignoredPaths.has(path)) {
      continue
    }

    const token = replaceHandleValue(
      path,
      {
        ...options,
        needEscaped: path.isStringLiteral() ? options.needEscaped ?? true : false,
      },
    )

    if (token) {
      replacementTokens.push(token)
    }
  }

  if (options.moduleSpecifierReplacements) {
    replacementTokens.push(
      ...collectModuleSpecifierReplacementTokens(analysis, options.moduleSpecifierReplacements),
    )
  }

  jsTokenUpdater
    .push(...replacementTokens)
    .filter(token => !ignoredPaths.has(token.path))
    .updateMagicString(ms)
  return ms
}

export function jsHandler(rawSource: string, options: IJsHandlerOptions): JsHandlerResult {
  let ast: ParseResult<File>
  try {
    ast = babelParse(rawSource, options.babelParserOptions)
  }
  catch (error) {
    return {
      code: rawSource,
      error: error as ParseError,
    } as JsHandlerResult
  }
  const analysis = analyzeSource(ast, options, jsHandler)
  const ms = processUpdatedSource(rawSource, options, analysis)

  const result: JsHandlerResult = {
    code: ms.toString(),
    get map() {
      return ms.generateMap()
    },
  }

  if (options.moduleGraph && options.filename) {
    const graph = new JsModuleGraph(
      {
        filename: options.filename,
        source: rawSource,
        analysis,
        handlerOptions: options,
      },
      options.moduleGraph,
    )

    const linked = graph.build()
    if (Object.keys(linked).length > 0) {
      result.linked = linked
    }
  }

  return result
}
