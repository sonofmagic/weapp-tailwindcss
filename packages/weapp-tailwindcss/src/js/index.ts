import type { CreateJsHandlerOptions, IJsHandlerOptions, JsHandler } from '../types'
import { defuOverrideArray } from '../utils'
import { jsHandler } from './babel'

export {
  jsHandler,
}

export function createJsHandler(options: CreateJsHandlerOptions): JsHandler {
  // Persist immutable options so repeated invocations only supply per-call overrides.
  const {
    mangleContext,
    arbitraryValues,
    escapeMap,
    jsPreserveClass,
    generateMap,
    babelParserOptions,
    ignoreCallExpressionIdentifiers,
    ignoreTaggedTemplateExpressionIdentifiers,
    uniAppX,
  } = options

  function handler(rawSource: string, classNameSet?: Set<string>, options?: CreateJsHandlerOptions) {
    const overrideOptions = (options ?? {}) as IJsHandlerOptions
    const resolvedOptions = defuOverrideArray<IJsHandlerOptions, IJsHandlerOptions[]>(
      {
        ...overrideOptions,
        classNameSet,
      },
      {
        classNameSet,
        escapeMap,
        arbitraryValues,
        mangleContext,
        jsPreserveClass,
        generateMap,
        babelParserOptions,
        ignoreCallExpressionIdentifiers,
        ignoreTaggedTemplateExpressionIdentifiers,
        uniAppX,
      },
    )

    return jsHandler(rawSource, resolvedOptions)
  }

  return handler
}
