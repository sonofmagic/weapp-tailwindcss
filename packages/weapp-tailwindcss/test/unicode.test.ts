import { createGetCase, unicodeCasePath } from './util'
import { decodeUnicode } from '@/js/handlers'
import { jsHandler, jsHandlerAsync } from '@/js'

const getCase = createGetCase(unicodeCasePath)

describe('unicode', () => {
  it('unicode case 0', () => {
    const testCase = '\u0009\u000A\u000B\u000C\u000D\u0020\u00A0\u1680\u2000\u2001\u2002'
      + '\u2003\u2004\u2005\u2006\u2007\u2008\u2009\u200A\u202F\u205F\u3000\u2028\u2029\uFEFF'
    expect(decodeUnicode(testCase)).toMatchSnapshot()
  })

  it('unicode case 0 fs', async () => {
    const testCase = await getCase('case0.js')

    const { code } = await jsHandlerAsync(testCase, {
      jsAstTool: 'ast-grep',
    })
    expect(code).toMatchSnapshot()
  })

  it('unicode case 1 fs', async () => {
    const testCase = await getCase('case0.js')

    const { code } = await jsHandler(testCase, {

    })
    expect(code).toMatchSnapshot()
  })

  it('unicode case 2 fs no replace', async () => {
    const testCase = await getCase('case2.js')
    const set: Set<string> = new Set()
    // set.add('after:content-[\'我知道我心,永恒12we_ds\']')
    // const x = 'after:content-[\'我知道我心,永恒12we_ds\']'
    // const y = 'after:content-[\'\\u6211\\u77e5\\u9053\\u6211\\u5fc3,\\u6c38\\u605212we_ds\']'
    const { code } = await jsHandlerAsync(testCase, {
      classNameSet: set,
    })
    expect(code).toMatchSnapshot()
  })

  it('unicode case 2 fs', async () => {
    const testCase = await getCase('case2.js')
    const set: Set<string> = new Set()
    set.add('after:content-[\'我知道我心,永恒12we_ds\']')

    const { code } = await jsHandlerAsync(testCase, {
      classNameSet: set,
    })
    expect(code).toMatchSnapshot()
  })

  it('unicode case 2 fs babel', async () => {
    const testCase = await getCase('case2.js')
    const set: Set<string> = new Set()
    set.add('after:content-[\'我知道我心,永恒12we_ds\']')
    const { code } = await jsHandler(testCase, {
      classNameSet: set,
    })
    expect(code).toMatchSnapshot()
  })
})
