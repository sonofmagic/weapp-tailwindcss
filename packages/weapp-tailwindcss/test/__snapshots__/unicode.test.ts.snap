// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`unicode > unicode case 0 1`] = `
"	

                　  ﻿"
`;

exports[`unicode > unicode case 0 fs 1`] = `
"const testCase = '\\u0009\\u000A\\u000B\\u000C\\u000D\\u0020\\u00A0\\u1680\\u2000\\u2001\\u2002'
      + '\\u2003\\u2004\\u2005\\u2006\\u2007\\u2008\\u2009\\u200A\\u202F\\u205F\\u3000\\u2028\\u2029\\uFEFF'"
`;

exports[`unicode > unicode case 1 fs 1`] = `
"const testCase = '\\u0009\\u000A\\u000B\\u000C\\u000D\\u0020\\u00A0\\u1680\\u2000\\u2001\\u2002'
      + '\\u2003\\u2004\\u2005\\u2006\\u2007\\u2008\\u2009\\u200A\\u202F\\u205F\\u3000\\u2028\\u2029\\uFEFF'"
`;

exports[`unicode > unicode case 2 fs 1`] = `"const x = "afterccontent-_qu6211u77e5u9053u6211u5fc3bu6c38u605212we_dsq_""`;

exports[`unicode > unicode case 2 fs babel 1`] = `"const x = "afterccontent-_qu6211u77e5u9053u6211u5fc3bu6c38u605212we_dsq_""`;

exports[`unicode > unicode case 2 fs no replace 1`] = `"const x = "after:content-['\\u6211\\u77e5\\u9053\\u6211\\u5fc3,\\u6c38\\u605212we_ds']""`;
